Mathlib/RingTheory/AdicCompletion/Functoriality.lean

```diff
@@ -0,0 +1,352 @@
+    (f.reduceModIdeal I) (Submodule.Quotient.mk (p := (I • ⊤ : Submodule R M)) x) =
+      Submodule.Quotient.mk (p := (I • ⊤ : Submodule R N)) (f x) :=
+  rfl
+
+end LinearMap
+
+namespace AdicCompletion
+
+open LinearMap
+
+theorem transitionMap_comp_reduceModIdeal (f : M →ₗ[R] N) {m n : ℕ}
+    (hmn : m ≤ n) : transitionMap I N hmn ∘ₗ f.reduceModIdeal (I ^ n) =
+      (f.reduceModIdeal (I ^ m) : _ →ₗ[R] _) ∘ₗ transitionMap I M hmn := by
+  ext x
+  simp
+
+namespace AdicCauchySequence
+
+/-- A linear map induces a linear map on adic cauchy sequences. -/
+@[simps]
+def map (f : M →ₗ[R] N) : AdicCauchySequence I M →ₗ[R] AdicCauchySequence I N where
+  toFun a := ⟨fun n ↦ f (a n), fun {m n} hmn ↦ by
+    have hm : Submodule.map f (I ^ m • ⊤ : Submodule R M) ≤ (I ^ m • ⊤ : Submodule R N) := by
+      rw [Submodule.map_smul'']
+      exact smul_mono_right _ le_top
+    apply SModEq.mono hm
+    apply SModEq.map (a.property hmn) f⟩
+  map_add' a b := by ext n; simp
+  map_smul' r a := by ext n; simp
+
+variable (M) in
+@[simp]
+theorem map_id : map I (LinearMap.id (M := M)) = LinearMap.id :=
+  rfl
+
+theorem map_comp (f : M →ₗ[R] N) (g : N →ₗ[R] P) :
+    map I g ∘ₗ map I f = map I (g ∘ₗ f) :=
+  rfl
+
+theorem map_comp_apply (f : M →ₗ[R] N) (g : N →ₗ[R] P) (a : AdicCauchySequence I M) :
+    map I g (map I f a) = map I (g ∘ₗ f) a :=
+  rfl
+
+end AdicCauchySequence
+
+/-- `R`-linear version of `adicCompletion`. -/
+private def adicCompletionAux (f : M →ₗ[R] N) :
+    AdicCompletion I M →ₗ[R] AdicCompletion I N :=
+  AdicCompletion.lift I (fun n ↦ reduceModIdeal (I ^ n) f ∘ₗ AdicCompletion.eval I M n)
+    (fun {m n} hmn ↦ by rw [← comp_assoc, AdicCompletion.transitionMap_comp_reduceModIdeal,
+        comp_assoc, transitionMap_comp_eval])
+
+@[local simp]
+private theorem adicCompletionAux_val_apply (f : M →ₗ[R] N) {n : ℕ} (x : AdicCompletion I M) :
+    (adicCompletionAux I f x).val n = f.reduceModIdeal (I ^ n) (x.val n) :=
+  rfl
+
+/-- A linear map induces a map on adic completions. -/
+def map (f : M →ₗ[R] N) :
+    AdicCompletion I M →ₗ[AdicCompletion I R] AdicCompletion I N where
+  toFun := adicCompletionAux I f
+  map_add' := by aesop
+  map_smul' r x := by
+    ext n
+    simp only [adicCompletionAux_val_apply, smul_eval, smul_eq_mul, RingHom.id_apply]
+    rw [val_smul_eq_evalₐ_smul, val_smul_eq_evalₐ_smul, map_smul]
+
+@[simp]
+theorem map_val_apply (f : M →ₗ[R] N) {n : ℕ} (x : AdicCompletion I M) :
+    (map I f x).val n = f.reduceModIdeal (I ^ n) (x.val n) :=
+  rfl
+
+/-- Equality of maps out of an adic completion can be checked on Cauchy sequences. -/
+@[ext]
+theorem map_ext {f g : AdicCompletion I M → N}
+    (h : ∀ (a : AdicCauchySequence I M),
+      f (AdicCompletion.mk I M a) = g (AdicCompletion.mk I M a)) :
+    f = g := by
+  ext x
+  apply induction_on I M x (fun a ↦ h a)
+
+/-- Equality of linear maps out of an adic completion can be checked on Cauchy sequences. -/
+theorem map_ext' {f g : AdicCompletion I M →ₗ[R] N}
```

Comment:
This should be the `@[ext]` lemma instead. See [partially-applied ext lemmas].

Reply:
Instead of `map_ext`?

Reply:
Yes.

Reply:
Okay maybe it doesn't matter.

Reply:
I swapped the `@[ext]`.